### ISSUE TITLE
Memory leak during inference

### ISSUE THREAD

**Issue by singhay** [axolotl-ai-cloud/axolotl#1181]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

The GPU memory should stay constant aka stateless w.r.t. the number of samples of inference increase.

### Current behaviour

The GPU memory keeps increasing as the number of samples of inference increase.

### Steps to reproduce

I could be wrong and wanted to get others eyes on it. 

I copy pasted the `do_inference()` method in cli>__init__.py into a new `do_eval()` method that does predictions in batch. I even turned off advanced sampling and did greedy decoding by setting `do_sample=False`. Even after that, the GPU memory (A100-80g) shoots very high to that point that after 3 data points I get a OOM error from GPU.

Surprisingly, the same code when run in `do_inference` mode, only take 11g memory and is stateless.

My dataset is of alpaca format and I adapted a base LLaMa model using config mentioned below. Kindly give it a look in case I'm doing something blatantly wrong. Thanks again for the awesome library!

```python
def do_eval(
    *,
    cfg: DictDefault,
    cli_args: TrainerCliArgs,
):
    model, tokenizer = load_model_and_tokenizer(cfg=cfg, cli_args=cli_args)
    default_tokens = {"unk_token": "<unk>", "bos_token": "<s>", "eos_token": "</s>"}

    for token, symbol in default_tokens.items():
        # If the token isn't already specified in the config, add it
        if not (cfg.special_tokens and token in cfg.special_tokens):
            tokenizer.add_special_tokens({token: symbol})

    prompter_module = AlpacaPrompter()

    model = model.to(cfg.device, dtype=cfg.torch_dtype)
    model.eval()

    generation_config = GenerationConfig(
        repetition_penalty=1.1,
        max_new_tokens=1024,
        temperature=0.9,
        top_p=0.95,
        bos_token_id=tokenizer.bos_token_id,
        eos_token_id=tokenizer.eos_token_id,
        pad_token_id=tokenizer.pad_token_id,
        # top_k=40,
        do_sample=False,
        use_cache=True,
        return_dict_in_generate=True,
        output_attentions=False,
        output_hidden_states=False,
        output_scores=False,
    )

    with open("/home/data/CDR/pred_test.jsonl", "w") as pf:
        with open("/home/asingh/data/CDR/test.jsonl") as f:
            with torch.no_grad():
                for l in tqdm(f.readlines(), total=22385):
                    record = json.loads(l)
                    prompt = prompter_module.build_prompt(instruction=record['instruction'],
                                                        input=record['input'])
                    batch = tokenizer(next(prompt), return_tensors="pt", add_special_tokens=True)
                    generated = model.generate(
                        inputs=batch["input_ids"].to(cfg.device),
                        generation_config=generation_config
                    )
                    record['prediction'] = tokenizer.decode(generated["sequences"].cpu().tolist()[0])
                    pf.write(json.dumps(record)+'\n')
```

### Config yaml

```yaml
base_model: NousResearch/Llama-2-7b-hf
model_type: LlamaForCausalLM
tokenizer_type: LlamaTokenizer
is_llama_derived_model: true

load_in_8bit: true
load_in_4bit: false
strict: false

datasets:
  - path: /home/data/CDR/train.jsonl
    ds_type: json
    type: alpaca
dataset_prepared_path:
val_set_size: 0.05
output_dir: ./lora-out

sequence_len: 4096
sample_packing: true
pad_to_sequence_len: true

adapter: lora
lora_model_dir:
lora_r: 32
lora_alpha: 16
lora_dropout: 0.05
lora_target_linear: true
lora_fan_in_fan_out:

wandb_project:
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model:

gradient_accumulation_steps: 4
micro_batch_size: 16
num_epochs: 1
optimizer: adamw_bnb_8bit
lr_scheduler: cosine
learning_rate: 0.0002

train_on_inputs: false
group_by_length: false
bf16: true
fp16: false
tf32: false

gradient_checkpointing: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 10
evals_per_epoch: 8
eval_table_size:
eval_table_max_new_tokens: 128
saves_per_epoch: 10
debug:
deepspeed:
weight_decay: 0.0
fsdp:
fsdp_config:
special_tokens:
  bos_token: "<s>"
  eos_token: "</s>"
  unk_token: "<unk>"
```

### Possible solution
I've tried detaching the input tensor, however, wasn't fruitful
```python
                    inputs = batch["input_ids"].to(cfg.device)
                    generated = model.generate(
                        inputs=inputs,
                        generation_config=generation_config
                    )
                    inputs.detach().cpu()
```

_No response_

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10

### axolotl branch-commit

main

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by winglian** [comment]
This is unlikely to be prioritized to be fixed. Please feel free to submit a PR if you track down the bug. Axolotl is primarily a training platform, and the provided inference solution is just a quick and dirty way to inference the trained model.
